RAII resource handles for CUDA objects

Andy-Jost opened the pull request:
## Summary

This PR introduces C++ RAII-based resource handles for CUDA objects in cuda.core. The handles use `std::shared_ptr` with custom deleters to ensure proper destruction order and prevent resource leaks.

For a detailed explanation of the design, see [`cuda_core/cuda/core/_cpp/DESIGN.md`](https://github.com/NVIDIA/cuda-python/pull/1368/files#diff-b8d0b7e0e8e0e8e0e8e0e8e0e8e0e8e0). (Tip: Select **⋯ → View file** to see the formatted Markdown.)

### Handle Types

- **ContextHandle**: Wraps `CUcontext` with proper retain/release semantics for primary contexts
- **StreamHandle**: Wraps `CUstream`, holds owning reference to its context
- **EventHandle**: Wraps `CUevent`, holds owning reference to its context  
- **MemoryPoolHandle**: Wraps `CUmemoryPool`, includes nvbug 5698116 workaround (clears peer access before destruction)
- **DevicePtrHandle**: Wraps `CUdeviceptr`, holds owning reference to memory pool (if applicable), includes IPC pointer cache

### Driver Workarounds

- **nvbug 5570902**: The IPC pointer cache works around a driver bug where multiple imports of the same device pointer via IPC are not reference-counted correctly.
- **Pinned memory IPC**: The cache also enables multiple imports of pinned memory buffers, working around `CUDA_ERROR_ALREADY_MAPPED` that the driver returns on duplicate imports.

### Benefits

- Eliminates noisy tracebacks from incorrect destruction order (see [#1303 comments](https://github.com/NVIDIA/cuda-python/issues/1303#issuecomment-2631556085))
- Enables safe resource ownership transfer between Python and C++
- Resources remain valid independent of Python garbage collection timing

## Related Issues

- Closes #1303 
- Closes #1342
- Related to #1187

## Test Plan

- [x] All existing tests pass
- [x] Pre-commit checks pass
- [x] New test for IPC duplicate import (nvbug 5570902)
- [x] Manual verification of destruction order fixes

Andy-Jost commented:
/ok to test 5576353b9bd0d1239278360981fde527f6d0c721

Andy-Jost commented:
I found a likely root-cause for the CI failures: ImportError due to circular import in cuda.core.experimental.__init__ (from cuda.core.experimental import _resource_handles). Proposed fix: use relative import (from . import _resource_handles) to avoid partially-initialized package import issues.

Andy-Jost commented:
/ok to test ccf9a3b217242773b9e9996a6ca1825b7e1140dc

Andy-Jost commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,877 @@
+}
+
+// Thread-local cache of primary contexts indexed by device ID
+thread_local std::vector<ContextHandle> primary_context_cache;
```
Primary context cache now lives in C++ (still thread local)

Andy-Jost commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,877 @@
+
+// Helper to clear peer access before destroying a memory pool.
+// Works around nvbug 5698116: recycled pool handles inherit peer access state.
+static void clear_mempool_peer_access(CUmemoryPool pool) {
```
This detail of memory pool lifetime management is removed from Cython. The Cython code can focus on implementing API features.

rwgk commented:
I used Cursor to discuss naming alternatives for the new family of `*Handle` types. The result is that I believe **`RcStream`**, **`RcContext`**, **`RcEvent`**, **`RcMemoryPool`**, **`RcDevicePtr`** would be a great choice.

## Why I think `Rc*` Is a Good Fit

The `Rc` prefix (short for "reference counted") accurately describes what these types are:

- **Explicit about mechanism**: These are `std::shared_ptr` wrappers with reference counting semantics. `Rc` communicates this directly.
- **Established precedent**: Rust's `Rc<T>` is widely recognized for the same concept (shared ownership via reference counting). Borrowing this convention leverages existing developer familiarity without conflicting with anything in the CUDA ecosystem.
- **Visually distinct type family**: The `Rc` prefix makes these types easy to grep for and immediately recognizable as a coherent family.
- **Factory function alignment**: `make_rc_stream()` or `create_rc_stream()` parallels the familiar `std::make_shared<T>()` pattern.

## Why I think `*Handle` Is Problematic

In CUDA and systems programming, "handle" typically means a **raw, lightweight, opaque identifier**—think `CUstream`, `CUcontext`, Windows `HANDLE`, or file descriptors. The common expectation:

| "Handle" implies | These types actually are |
|------------------|-------------------------|
| Lightweight, raw | `shared_ptr` with refcount overhead |
| Non-owning | Manage ownership via custom deleters |
| No lifetime semantics | Encode structural dependencies (stream → context) |

Calling these `StreamHandle` when `CUstream` is *also* called a "handle" creates confusion. A developer might reasonably assume `StreamHandle` is a typedef for `CUstream` or a thin wrapper, not a reference-counted smart pointer.

## Alternatives Considered

| Name | Pros | Cons |
|------|------|------|
| `StreamPtr` | Familiar smart-pointer convention | Could still be confused with raw pointers |
| `SharedStream` | Mirrors `shared_ptr` | Verbose; "Shared" prefix is overloaded |
| `StreamLife` | Emphasizes lifetime management | Unusual; might take getting used to |
| `StreamLease` | Nice metaphor for shared ownership | Implies time-bounded or revocable |
| `StreamProxy` | Intermediary pattern | Emphasizes indirection, not ownership |
| `StreamHolder` / `StreamKeep` | Descriptive | Less precise about mechanism |

## My Conclusion

`Rc*` strikes the right balance: it's **concise**, **precise about semantics**, **established elsewhere** (Rust), and **not already overloaded** in the CUDA context. I think it would serve developers well as they learn and use this new abstraction layer.

Andy-Jost commented on cuda_core/cuda/core/_cpp/resource_handles.hpp:
```diff
@@ -0,0 +1,298 @@
+    return h ? static_cast<std::intptr_t>(*h) : 0;
+}
+
+// py() - convert handle to Python driver wrapper object
```
I can refactor to something like `as_py_impl`. I don't think any synchronization is needed here. At best it can avoid the rare case where `cls` initialization happens more than once, but that is harmless.

rwgk replied:
> I don't think any synchronization is needed here.

Did you see that the free-threading case has UB (Summary table above)?

I recommend not underestimating race issues. E.g. I had a very time-consuming run-in with that while working on pybind11 (even before free-threading was a thing):

https://github.com/pybind/pybind11/pull/4877

Note that the racing code "worked" for years, until numpy 2 came around, and for reasons I still don't understand, suddenly it became very flaky.

My own opinion:

It's so easy to make this right now (LLMs). I'd adopt the already generated Option B, and probably we'll never have to think about this again.

rwgk commented:
> it seems AI generated and doesn't really reflect reality

It's a good mix of my own thoughts and what Cursor suggested. I probably spent almost an hour discussing the naming with Cursor. The final formatting is almost pure Cursor (I deleted some stuff).

> From my point of view the Rc analogy doesn't hold much water.

Why I picked up Rc: ref-counted is fundamental to the new type family. The Rust analogy is just nice. Beyond discussions on this PR, we don't have to make any reference to the Rust analogy; but conveniently, when a Rust person looks at this code, their intuition will match.

Also on my mind: that `shared_ptr` is being used is an implementation detail, more private than public, although I think it's most pragmatic not to try to hide it.

Andy-Jost commented:
/ok to test 4e54ca6ae1c5bef96bd148278db59735f84f7c99

rwgk commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,877 @@
+
+struct DevicePtrBox {
+    CUdeviceptr resource;
+    mutable StreamHandle h_stream;
```
Suggested comment for `h_stream`:

```c++
    // Mutable to allow set_deallocation_stream() to update the stream
    // through a const DevicePtrHandle. The stream can be changed after
    // allocation (e.g., to synchronize deallocation with a different stream).
```

rwgk commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,877 @@
+namespace {
+
+static std::once_flag driver_load_once;
+static bool driver_loaded = false;
```
I believe `static` is redundant inside an anonymous namespace. Harmless but I'd remove it.

rwgk reviewed (approved):
Some additional pretty minor suggestions. Technically this PR looks great to me. My main concern is intuitive naming, I believe `Holder` is far better than `Handle`, for the reasons explained under **"Handle" implies** in [my comment from a couple days ago](https://github.com/NVIDIA/cuda-python/pull/1368#issuecomment-3712495521).

rparolin commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,877 @@
+StreamHandle create_stream_handle_with_owner(CUstream stream, PyObject* owner) {
+    Py_XINCREF(owner);
+    auto box = std::shared_ptr<const StreamBox>(
+        new StreamBox{stream},
```
`std::make_shared` will clean up the `StreamBox` allocation if an exception is thrown during construction.

Andy-Jost commented:
> AFAICT this is probably fine for free-threaded Python, but it has the exact same latent deadlock risk in multi-threaded scenarios that we had to solve with https://github.com/pybind/pybind11/pull/4877.

Thanks for pointing this out. This is a pretty subtle issue that could cause a headache later. I dealt with this in the latest upload as follows:
- Removed the static class lookup in `py` functions. The deadlock risk is real and I don't think the extra complexity is justified for these functions. They are just nice-to-haves, not critical to performance.
- I found an additional instance of this problem related to `load_driver_api` and updated it to release the GIL before entering `std::call_once`.
- I updated the DESIGN.md document to point out the deadlock hazard.

As for the naming, let's put it up to a vote. I have a slight preference for the existing names, but I don't mind changing the code if others prefer the new names. I mainly want to avoid changing multiple times. There are two proposed name changes:

1. `cu`/`py`/`intptr` -> `as_cu`/`as_py`/`as_intptr`
2. `StreamHandle` -> `StreamHolder` etc.

The second one also involves renaming files, documentation, etc. If anyone has a preference either way, please mention it.

I will get to work on the remaining comments now.

leofang commented on cuda_core/cuda/core/_cpp/DESIGN.md:
Thanks for writing up the markdown file, Andy. But the added complexity, especially around the new Cython bindings (to the resource handle C++ code),  and the GIL handling, warrants at least another discussion, I think. This complexity was not mentioned in the original design docs (or maybe I missed it?), and I am not sure it is the best way to go. I'll schedule a meeting to follow up.

Andy-Jost replied:
As discussed in the meeting, I'll follow up to see whether the capsules can be eliminated. See #1450 and #1452

leofang commented on cuda_core/cuda/core/experimental/_context.pxd:
I don't think this file is needed

Andy-Jost replied:
Oh, this is under `experimental/` -- yeah not needed at all

Andy-Jost commented on cuda_core/cuda/core/_cpp/resource_handles.hpp:
```diff
@@ -0,0 +1,266 @@
+// ============================================================================
+
+// cu() - extract the raw CUDA handle
+inline CUcontext cu(const ContextHandle& h) noexcept {
```
As a general rule, yes, short names should be avoided. In this case, `cu()` and `py()` are ubiquitous operations used throughout the codebase, and longer names would arguably harm readability.

These functions are designed to be nested in expressions, so brevity matters.

For example:

```
HANDLE_RETURN(cydriver.cuMemcpyAsync(
   cu(dst._h_ptr), cu(self._h_ptr), src_size, cu(s._h_stream)))
```

With a longer name like `get_cuda_handle()`:

```
HANDLE_RETURN(cydriver.cuMemcpyAsync(
    get_cuda_handle(dst._h_ptr), get_cuda_handle(self._h_ptr),
    src_size, get_cuda_handle(s._h_stream)))
```

The longer form requires more lines, more visual parsing, and potentially more temporary variables. Given how frequently these appear in driver call expressions, I believe the short names reduce cognitive load rather than increase it.

Andy-Jost replied:
If we're trying to push this PR through today, I suggest we table this question and follow up, if needed.

Andy-Jost commented on cuda_core/cuda/core/_cpp/resource_handles.cpp:
```diff
@@ -0,0 +1,916 @@
+
+void clear_last_error() noexcept {
+    err = CUDA_SUCCESS;
+}
```
I think this is another question we should follow up on, if needed.

Andy-Jost commented:
/ok to test b3c6c648b888eeb22f129654987ff9ccdb5d6ce6

Andy-Jost commented:
The latest upload mass-renames `py`, `cu`, and `intptr` to prefix `as_`.
